src/NewRemoting/MessageHandler.cs

```diff
@@ -41,10 +59,215 @@ public MessageHandler(InstanceManager instanceManager, FormatterFactory formatte
+									proxyType = typeof(DelegateProxyOnClient<>).MakeGenericType(arguments.ToArray());
+									break;
+								case 2:
+									proxyType = typeof(DelegateProxyOnClient<,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 3:
+									proxyType = typeof(DelegateProxyOnClient<,,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 4:
+									proxyType = typeof(DelegateProxyOnClient<,,,>).MakeGenericType(arguments.ToArray());
+									break;
+								default:
+									throw new NotSupportedException(
+										$"Unsupported number of arguments for action ({arguments.Count}");
+							}
+						}
+
+						var proxy = Activator.CreateInstance(proxyType);
+						var addEventMethod = proxyType.GetMethod("add_Event");
+						addEventMethod.Invoke(proxy, new[] { del });
+						_instanceManager.AddInstance(proxy, instanceId, referencesWillBeSentTo, proxyType);
+
+						w.Write((int)RemotingReferenceType.AddEvent);
+						LogMsg(RemotingReferenceType.AddEvent);
+						w.Write(instanceId);
+						w.Write(hasReturnValue);
+						// If the type of the delegate method is generic, we need to provide its type arguments
+						w.Write(arguments.Count);
+						foreach (var argType in arguments)
+						{
+							string arg = argType.AssemblyQualifiedName;
+							if (arg == null)
+							{
+								throw new RemotingException("Unresolved generic type or some other undefined case");
+							}
+
+							w.Write(arg);
+						}
+					}
+				}
+				else
+				{
+					// The delegate target is a static method
+					////						w.Write(string.Empty);
+					throw new InvalidOperationException("The delegate target is a static method");
+				}
+			}
+			else if (calledMethod.IsSpecialName && calledMethod.Name.StartsWith("remove_", StringComparison.Ordinal))
+			{
+				if (del.Target != null)
+				{
+					string instanceId = _instanceManager.GetDelegateTargetIdentifier(del, remoteInstanceId);
+
+					if (_instanceManager.TryGetObjectFromId(instanceId, out var existingDelegateProxy))
+					{
+						// Create a proxy class that provides a matching event for our delegate
+						var proxyType = existingDelegateProxy.GetType();
+						var removeEventMethod = proxyType.GetMethod("remove_Event");
+						removeEventMethod.Invoke(existingDelegateProxy, new[] { del });
+						var isEmptyMethod = proxyType.GetMethod("IsEmpty");
+						bool isEmpty = (bool)isEmptyMethod.Invoke(existingDelegateProxy, null);
+						if (isEmpty)
+						{
+							_instanceManager.Remove(instanceId, _instanceManager.InstanceIdentifier);
+						}
+						else
+						{
+							// proxy is still needed, whole message can be aborted here
+							return false;
+						}
+					}
+
+					w.Write((int)RemotingReferenceType.RemoveEvent);
+					LogMsg(RemotingReferenceType.RemoveEvent);
+					w.Write(instanceId);
+				}
+				else
+				{
+					// The delegate target is a static method
+					throw new InvalidOperationException("The delegate target is a static method");
```

Comment:
The exception here is fine, but we should probably try to understand whether this is always redundant to the initial test at the beginning of the method.